src/cargo/core/compiler/build_context/target_info.rs

```diff
@@ -990,6 +993,35 @@ impl<'gctx> RustcTargetData<'gctx> {
             }
         };
 
+        let check_cfg = {
+            let mut process = rustc.workspace_process();
+
+            apply_env_config(gctx, &mut process)?;
+            process
+                .arg("-")
+                .arg("--print=check-cfg")
+                .arg("--check-cfg=cfg()")
+                .arg("-Zunstable-options")
+                .env_remove("RUSTC_LOG");
+
+            // Removes `FD_CLOEXEC` set by `jobserver::Client` to pass jobserver
+            // as environment variables specify.
+            if let Some(client) = gctx.jobserver_from_env() {
+                process.inherit_jobserver(client);
+            }
+
+            let (output, _error) = rustc
+                .cached_output(&process, 0)
+                .with_context(|| "failed to run `rustc` to learn about check-cfg information")?;
+
+            let lines = output.lines();
+
+            lines.fold(CheckCfg::default(), |mut check_cfg, line| {
+                check_cfg.process_line(line);
+                check_cfg
+            })
+        };
```

Comment:
> The feature voluntarily does not follow RUSTFLAGS (to avoid mistakes with one-off and for consistency) but instead follows Rust unexpected_cfgs lint configuration:

Could you clarify what you mean by this?

Calling rustc is "expensive" for no-op builds which is especially important for cargo-script.  Yes, this is cached, but I'd like to better understand why we need this over adding it to another call.

Reply:
I want to avoid confusing users who might add a `--check-cfg` args in `RUSTFLAGS`, thinking that it works while at the second they will remove them from `RUSTFLAGS` it will not anymore; but now that I think about it more it's already the case for `rustc`, so this would be worse as it would work with `rustc` but not Cargo.

I will revert this behaviour and merge it into the others `--print`s.

Reply:
Having this respect RUSTFLAGS would likely require this to be processed during compilation.  Properly processing RUSTFLAGS is a complex, expensive operation as it is fixed point as it evaluates how new `--cfg`s opt-in more RUSTFLAGS.

Reply:
`TargetInfo` as some handling of `RUSTFLAGS` so now use and it's handling to have `RUSTFLAGS`. [`18f2caf` (#14581)](https://github.com/rust-lang/cargo/pull/14581/commits/18f2caf13dd37e95720d2d9b151d7af2d2f2e520)